File: tests/io_tests.rs
use lorealdb::db_engine::DBEngine;
use pyo3::prelude::*;
use std::time::{Duration, Instant};
use tempfile::tempdir;

#[test]
fn benchmark_background_io_throughput() {
    Python::initialize();
    let dir = tempdir().expect("Failed to create temp dir");
    let db_path = dir.path().join("io_benchmark.redb");
    let engine = DBEngine::new(db_path.to_str().unwrap()).expect("Failed to init engine");

    let total_records = 100_000;
    let mut data_store = Vec::with_capacity(total_records);
    let mut total_bytes = 0;

    // Generate realistic JSON payloads to trigger the simd_json indexing logic
    for i in 0..total_records {
        let id = format!("ticket_{}", i);
        let payload = format!(
            "{{\"status\": \"open\", \"priority\": \"high\", \"id\": \"{}\", \"description\": \"This is a simulated payload to test the I/O throughput and metadata indexing of the embedded database engine.\"}}",
            id
        );
        total_bytes += payload.len();
        data_store.push((id, payload));
    }

    println!("\n--- Starting Background I/O Write Benchmark ---");
    println!("Total Records:      {}", total_records);
    println!(
        "Total Payload Size: {:.2} MB",
        total_bytes as f64 / 1_048_576.0
    );

    // system_start measures from the moment we start pushing to the queue
    // until the data is fully written to the disk.
    let system_start = Instant::now();

    // 1. Measure Queue Time (Main Thread blocking time)
    let queue_start = Instant::now();
    Python::attach(|py| {
        engine
            .insert_many_json(py, data_store)
            .expect("Batch insert failed");
    });
    let queue_time = queue_start.elapsed();

    // 2. Measure Background Commit Time (Disk I/O and Indexing)
    let commit_start = Instant::now();
    let last_id = format!("ticket_{}", total_records - 1);

    // Poll the DB until the very last record appears
    loop {
        let is_committed = Python::attach(|py| engine.get(py, &last_id).unwrap().is_some());

        if is_committed {
            break;
        }
        // Yield the CPU briefly to prevent starving the background worker thread
        std::thread::sleep(Duration::from_millis(2));
    }

    let commit_wait_time = commit_start.elapsed();
    let total_system_time = system_start.elapsed();

    // 3. Calculate Performance Metrics
    let records_per_sec = total_records as f64 / total_system_time.as_secs_f64();
    let mb_per_sec = (total_bytes as f64 / 1_048_576.0) / total_system_time.as_secs_f64();

    println!("-----------------------------------------------");
    println!("Queue Time (Main Thread):   {:?}", queue_time);
    println!("Wait Time (Background I/O): {:?}", commit_wait_time);
    println!("Total System Write Time:    {:?}", total_system_time);
    println!("-----------------------------------------------");
    println!("Throughput (Records/sec):   {:.0} ops/sec", records_per_sec);
    println!("Throughput (MB/sec):        {:.2} MB/s", mb_per_sec);
    println!("-----------------------------------------------\n");
}

File: benchmarks/io_tests.py
import os
import tempfile
import time

from lorealdb import DBEngine


def benchmark_background_io_throughput():
    # Use tempfile to replicate tempdir()
    with tempfile.TemporaryDirectory() as temp_dir:
        db_path = os.path.join(temp_dir, "io_benchmark.redb")
        engine = DBEngine(db_path)

        total_records = 100_000
        data_store = []
        total_bytes = 0

        # Generate realistic JSON payloads
        for i in range(total_records):
            id_str = f"ticket_{i}"
            payload = (
                f'{{"status": "open", "priority": "high", "id": "{id_str}", '
                f'"description": "This is a simulated payload to test the I/O '
                f'throughput and metadata indexing of the embedded database engine."}}'
            )
            # Encode to utf-8 to get the exact byte length for MB/s calculation
            total_bytes += len(payload.encode("utf-8"))
            data_store.append((id_str, payload))

        print("\n--- Starting Background I/O Write Benchmark ---")
        print(f"Total Records:      {total_records}")
        print(f"Total Payload Size: {total_bytes / 1_048_576.0:.2f} MB")

        system_start = time.perf_counter()

        # 1. Measure Queue Time (Main Thread blocking time)
        queue_start = time.perf_counter()

        # insert_many_json takes a list of (id, json_string) tuples
        engine.insert_many_json(data_store)

        queue_time = time.perf_counter() - queue_start

        # 2. Measure Background Commit Time (Disk I/O and Indexing)
        commit_start = time.perf_counter()
        last_id = f"ticket_{total_records - 1}"

        # Poll the DB until the very last record appears
        while True:
            if engine.get(last_id) is not None:
                break
            # Yield the CPU briefly (2ms) to prevent starving the background worker thread
            time.sleep(0.002)

        commit_wait_time = time.perf_counter() - commit_start
        total_system_time = time.perf_counter() - system_start

        # 3. Calculate Performance Metrics
        records_per_sec = total_records / total_system_time
        mb_per_sec = (total_bytes / 1_048_576.0) / total_system_time

        print("-----------------------------------------------")
        # Format to 4 decimal places for milliseconds readability
        print(f"Queue Time (Main Thread):   {queue_time:.4f}s")
        print(f"Wait Time (Background I/O): {commit_wait_time:.4f}s")
        print(f"Total System Write Time:    {total_system_time:.4f}s")
        print("-----------------------------------------------")
        print(f"Throughput (Records/sec):   {records_per_sec:.0f} ops/sec")
        print(f"Throughput (MB/sec):        {mb_per_sec:.2f} MB/s")
        print("-----------------------------------------------\n")


if __name__ == "__main__":
    benchmark_background_io_throughput()

# Results for reference.
# python io_tests.py

# --- Starting Background I/O Write Benchmark ---
# Total Records:      100000
# Total Payload Size: 17.92 MB
# -----------------------------------------------
# Queue Time (Main Thread):   2.4097s
# Wait Time (Background I/O): 0.5875s
# Total System Write Time:    2.9973s
# -----------------------------------------------
# Throughput (Records/sec):   33364 ops/sec
# Throughput (MB/sec):        5.98 MB/s
# -----------------------------------------------
